### PR TITLE
future proof open-ended ProtocolType

### PR DESCRIPTION
**Pull request by hbagdi**
See #390

### CODE REVIEW

**Comment by hbagdi** [comment]
This has largely been borrowed from [AppProtocol](https://github.com/kubernetes/kubernetes/blob/master/pkg/apis/discovery/types.go#L138-L145).

**Review by howardjohn — APPROVED** [review]

File: apis/v1alpha1/gateway_types.go [hunk 1]
```diff
@@ -298,6 +298,12 @@ const (
 // * "TCP"
 // * "UDP"
 //
+// Implementations can define their own protocols if one is not already defined.
+// Un-prefixed names are reserved for IANA standard service names (as per
+// RFC-6335 and http://www.iana.org/assignments/service-names).
+// Non-standard protocols should use prefixed names such as
+// mycompany.com/my-custom-protocol.
```
Comment (robscott) [inline apis/v1alpha1/gateway_types.go:305]: Do we need to clarify the support level here? Assuming this is custom?

Comment (hbagdi) [reply apis/v1alpha1/gateway_types.go:297]: Updated.

Comment (jpeach) [reply apis/v1alpha1/gateway_types.go:297]: I think that this needs a broader discussion and concept. The protocols we currently specify are spread across multiple OSI layers (e.g. we have both "UDP" and "HTTP"), and can be protocol families (e.g. "HTTPS" probably means HTTP/1.1 and HTTP/2, possibly HTTP/3).

The IANA service-names registry is just UDP/TCP port registrations. There's no "h2" or "GRPC" for example.

Comment (robscott) [reply apis/v1alpha1/gateway_types.go:297]: @jpeach Those are good points, do you know of any better lists we could use here?

Comment (danehans) [reply apis/v1alpha1/gateway_types.go:297]: Consider splitting `ProtocolType` into `AppProtocolType` that contains application layer protocols (i.e. HTTP) and `TransportProtocolType` that contains transport layer protocols (i.e. TCP).

**Comment by robscott** [comment]
@thockin Do you have any ideas on how we could expand the list of protocols we support on Gateway listeners?

File: apis/v1alpha1/gateway_types.go (hunk 1 above)
Comment (robscott) [reply apis/v1alpha1/gateway_types.go:297]: At least short term, what if we remove the RFC reference entirely and simplify the options to:
- Protocols we've listed and officially support
- Custom protocols can be specified with domain prefixes

We can always expand our officially supported list in the future but this feels like a reasonable compromise for now.

Comment (hbagdi) [reply apis/v1alpha1/gateway_types.go]: @robscott I like that approach. I've updated the PR to reflect that.

File: apis/v1alpha1/gateway_types.go [hunk 2]
```diff
@@ -295,15 +295,11 @@ const (
...
-// * "TLS"
-// * "TCP"
-// * "UDP"
-//
-// +kubebuilder:validation:Enum=HTTP;HTTPS;TLS;TCP;UDP
+// Implementations can define their own protocols if one is not already defined.
+// Un-prefixed names are reserved for protocols defined in this package.
+// Non-standard protocols should use prefixed names such as
+// mycompany.com/my-custom-protocol.
+// Any protocol defined by implementations will fall under custom conformance.
```
Comment (robscott) [inline apis/v1alpha1/gateway_types.go:302]: I think we probably still want to retain a list of core protocol types that we support. Maybe just replace `Valid ProtocolType values...` with `Core ProtocolType values...`. Then you can add the paragraph about custom protocols below that.

Comment (hbagdi) [reply apis/v1alpha1/gateway_types.go]: Reworded this a little bit. PTAL.

**Comment by robscott** [comment]
Thanks! Interested in any final feedback from @danehans or @jpeach here.

/approve

**Comment by danehans** [comment]
/lgtm

I created https://github.com/kubernetes-sigs/service-apis/issues/433 to capture the need to define "Extended" protocol types.

**Pull request merged**